File: 3660-jump-game-ix/3660-jump-game-ix.py
class Solution:
    def maxValue(self, nums: List[int]) -> List[int]:
        n=len(nums)
        premax=[0]*n
        suffmax=[0]*n
        ans=[0]*n
        premax[0]=nums[0]
        for i in range(1,n):
            premax[i]=max(premax[i-1],nums[i])
        suffmax[n-1]=nums[n-1]
        for i in range(n-2,-1,-1):
            suffmax[i]=min(suffmax[i+1],nums[i])
        ans[n-1]=premax[n-1]
        for i in range(n-2,-1,-1):
            if(premax[i]>suffmax[i+1]):
                ans[i]=ans[i+1]
            else:
                ans[i]=premax[i]
        return ans

File: 3660-jump-game-ix/3660-jump-game-ix.cpp
class Solution {
public:
    vector<int> maxValue(vector<int>& nums) {
        int n=nums.size();
        vector<int>prefixmax(n),suffixmax(n),ans(n);
        prefixmax[0]=nums[0];
        for(int i=1;i<n;i++){
            prefixmax[i]=max(prefixmax[i-1],nums[i]);
        }
        suffixmax[n-1]=nums[n-1];
        for(int i=n-2;i>=0;i--){
            suffixmax[i]=min(suffixmax[i+1],nums[i]);
        }
        ans[n-1]=prefixmax[n-1];
        for(int i=n-2;i>=0;i--){
            if(prefixmax[i]>suffixmax[i+1]) ans[i]=ans[i+1];
            else ans[i]=prefixmax[i];
        }
        return ans;
        
    }
};